### PR TITLE
Version 1

### PR DESCRIPTION
**Pull request by ymmt2005**
Feature list:
- Use [`cybozu-go/cmd`](https://github.com/cybozu-go/cmd) framework to:
  - Graceful stop & restart.
  - Logging options.
  - etc.
- Implement SOCKS4 and SOCKS4a
  - Drop dependency on `github.com/armon/go-socks5` as it is made only for SOCKS5.

### CODE REVIEW

**Comment by ymmt2005** [comment]
@tfukaya 
Added our own SOCKS library.
Although CI fails due to lint errors, the new library passes all tests.

Please take a look!

**Comment by ymmt2005** [comment]
It's ready! @tfukaya

File: socks/server_test.go [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+	}
+
+	curl = exec.Command("curl", "-4", "-I", "--socks5", addr, url1)
+	err = curl.Run()
+	if err == nil {
+		t.Error("SOCKS5 w/o hostname should be denied")
+	}
+
+	if hasIPv6() {
+		curl = exec.Command("curl", "-6", "-I", "--socks5-hostname", addr, url1)
```
Comment (ghost) [inline socks/server_test.go]: I think this test should be done for IPv4 too.

Comment (ymmt2005) [reply socks/server_test.go]: Fixed!

**Comment by ghost** [comment]
good to work, merged!

**Comment by ymmt2005** [comment]
Thank you!

**Pull request merged**